src/core/reactor_backend.cc

```diff
@@ -182,6 +182,15 @@ aio_storage_context::handle_aio_error(linux_abi::iocb* iocb, int ec) {
             // we will only remove it from _pending_io and try again.
             return 1;
         }
+        case ENOTSUP: {
```

Comment:
I guess `case EOPNOTSUPP` should also come here, I found this in ext4 code:

```
static ssize_t ext4_buffered_write_iter(struct kiocb *iocb,
                    struct iov_iter *from)
{
    ssize_t ret;
    struct inode *inode = file_inode(iocb->ki_filp);

    if (iocb->ki_flags & IOCB_NOWAIT)
        return -EOPNOTSUPP;
```

it's buffered, but ext4 sometimes falls back to it even if O_DIRECT is requested

Reply:
Also check aio_read() code:

```
    if (unlikely(!file->f_op->read_iter))
        return -EINVAL;
```

some filesystems may bail out early with EINVAL for any AIO

Reply:
> Also check aio_read() code:
> 
> ```
>     if (unlikely(!file->f_op->read_iter))
>         return -EINVAL;
> ```
> 
> some filesystems may bail out early with EINVAL for any AIO

yeah, if the aio read or write op is not implemented at all.